posthog/api/plugin.py

```diff
@@ -124,26 +127,89 @@ def destroy(self, request: request.Request, *args, **kwargs) -> Response:
 
 
 class PluginConfigSerializer(serializers.ModelSerializer):
+    config = serializers.SerializerMethodField()
+
     class Meta:
         model = PluginConfig
         fields = ["id", "plugin", "enabled", "order", "config", "error"]
         read_only_fields = ["id"]
 
+    def get_config(self, plugin_config: PluginConfig):
+        attachments = PluginAttachment.objects.filter(plugin_config=plugin_config).only(
+            "id", "file_size", "file_name", "content_type"
+        )
+        new_plugin_config = plugin_config.config.copy()
+        for attachment in attachments:
+            new_plugin_config[attachment.key] = {
+                "uid": attachment.id,
+                "saved": True,
+                "size": attachment.file_size,
+                "name": attachment.file_name,
+                "type": attachment.content_type,
+            }
+        return new_plugin_config
+
     def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> PluginConfig:
         if not can_configure_plugins_via_api():
             raise ValidationError("Plugin configuration via the web is disabled!")
         request = self.context["request"]
         validated_data["team"] = request.user.team
+        self._fix_formdata_config_json(validated_data)
         plugin_config = super().create(validated_data)
+        self._update_plugin_attachments(plugin_config)
         reload_plugins_on_workers()
         return plugin_config
 
     def update(self, plugin_config: PluginConfig, validated_data: Dict, *args: Any, **kwargs: Any) -> PluginConfig:  # type: ignore
```

Comment:
As a django `n00b` I'm not sure what to do here. 

> So this has some interesting implicit arguments - it relies on being called to update inside a request.

I have seen code like `request = self.context["request"]` in a lot of serializers, so I'm not sure if this is something that we should untangle... nor how:

![image](https://user-images.githubusercontent.com/53387/98956975-f5dc1100-2500-11eb-912b-14507f65f0c9.png)

Keep in mind, that I need to make sure a lot of different pieces work together here: 1) `rc-form` in antd and its `rc-upload` file upload component, both for adding new files and for displaying old files (custom metadata format) 2) a unified form for config vars that live in two models (plugin config and plugin attachments), 3) file uploads with FormData while also sending non-file fields, 4) JSON serialization for `config` vars, 5) receiving the file in django and storing it in postgresql with custom fields for metadata.

If you have some suggestions on how to clean this up into a more djangoesque style, while retaining all those assumptions, please let me know. I'm just happy I got anything to work :).

Reply:
Not too sure either how to do this without introducing an extra mapping layer to be honest.

A start would be exposing the request as argument on `_fix_formdata_config_json`, `_update_plugin_attachments` instead of it being read from context - making the thing a bit more pure and closer to testable. But that does relatively little on it's own.

Reply:
I'm not sure what would this solve and why do we need to make these `_internal` methods of the serializer pure and testable. Where would I test them?

From what I can guess, the entire point of OOP is that you have access to instance variables when you need them.

I have added a bunch of tests to the plugin API (file uploads still to come) that indirectly test these methods, so... 🤔 🤷